### ISSUE TITLE
What is the mAP expected to be when training from vgg16 checkpoint?

### ISSUE THREAD

**Issue by joaqo** [pierluigiferrari/ssd_keras#66]
Hey, if I understand correctly the weights linked in the README are the weights from the original caffe implementation converted to tensorflow, so that would mean they weren't trained using this particular implementation of the paper.

If thats the case, I was wondering what the expected mAP should be when trained using this implementation? I've been looking at this other implementation: https://github.com/balancap/SSD-Tensorflow and they get state of the art performance when loading the original caffe weights, but only .60 mAP when training their own weights on VOC 07+12. So I started wondering if this is a more widespread problem, and if anyone had managed to get the performance of the original implementation ~77 mAP.

Great implementation btw, really loved how everything was explained in great details in the comments!

**Comment by pierluigiferrari** [comment]
The VGG-16 weights that were linked in the README before were adapted from `keras.applications.vgg16` and I don't know from where they got them in turn. I believe I followed the sub-sampling and convolutionalization procedure of the Caffe implementation exactly, but I can't be absolutely certain.

Because of this, I've just replaced the VGG-16 weights with a direct port of the `.caffemodel` file that is provided in the repository of the original Caffe SSD implementation. Check out the new link. So now you can be certain that in case you are unable to reproduce the mAP of the original models in your training, at least the VGG-16 weights cannot be the reason - if they weren't already identical to the corresponding `.caffemodel` weights before, then they definitely are now.

That being said, I don't think the VGG-16 weights are the problem in the TensorFlow SSD implementation you linked above. I am not super familiar with that implementation, but I think it is very unlikely that a small potential difference in the trained VGG-16 weights could lead to a 17% mAP difference after the training converged. Such a potential small difference in the initialized weights should become almost irrelevant as the training progresses and, if anything, lead to a small mAP difference upon convergence. This is, however, just my intuition, I might be wrong.

I think it is a lot more likely that differences in the training process, the loss function, etc. are the reason for why they can't achieve the mAP score of the original implementation. There might be subtle differences in how the default boxes are matched, in the applied data augmentation, in how the loss is computed, or wherever else. Once again, I'm not super familiar with that TensorFlow implementation and this is all just speculation. But it doesn't seem plausible that the VGG-16 weights are solely responsible for a 17% mAP difference.

**Comment by joaqo** [comment]
Hey @pierluigiferrari thanks for the answer. I am afraid I may not have explained myself correctly. I wasn't talking about the VGG16 weights, I was talking about the weights for the whole model (VGG16 + SSD Multibox layers). In the Tensorflow-SSD implementation they seem to have created a checkpoint with all of the weights from the original caffe implementation, and from looking at the issues posted there, no one seems to be able to reproduce the performance of said weights.

So if I am understanding correctly, you just downloaded the VGG16 weights, and then fine tuned the whole network after. That is exactly what I want to do, do you mind asking what mAP you achieved doing that? Thanks!

**Comment by pierluigiferrari** [comment]
Oh, in that case I'm a bit confused by your question. The README clearly states:

> Here are the ported weights for all the original trained models.

I.e. the provided `.h5` files are direct ports of the original `.caffemodel` files.

The mAP values you see in the evaluation results section of the README are the exact results I received when I evaluated the respective models with those trained weights. I thought this was also clear, otherwise how would I be able to report any mAP numbers if I hadn't measured them myself? As the evaluation results section states:

> Here are the mAP evaluation results from the official Pascal VOC Matlab evaluation script (for 2007 test) and the official test server (for 2012 test). In all cases the results are either identical to those of the original Caffe models or surpass them slightly.

I.e. I ran the evaluation using the official Pascal VOC test server (for VOC2012 test) and Matlab evaluation script (for VOC2007 test), and what you see there are the results.

**Comment by joaqo** [comment]
I think we are not understanding each other. Its probably me not being clear with my question so sorry about that. Let me try to be clearer and more concise:

- In the Tensorflow-SSD implementation, the author ported the weights from the caffe implementation and got the same mAP as the paper. But when he, or the people reporting the issues, tried to train his implementation starting from just the VGG16 weights, he never could achieve the same results (he got .60mAP)
- I saw in your README, as you quoted here, that you also ported the weights from the caffe implementation, and also got the same mAP as the paper by doing so.
- What I was wondering is what mAP does your implementation get when trained just from the VGG16 weights, so I can compare it with the 0.60mAP that  the Tensorflow-SSD implementation got.

Thanks a lot and sorry again if I am not being clear enough.

**Comment by pierluigiferrari** [comment]
Sorry for the misunderstanding, in hindsight your initial question was clear, I don't know why I didn't understand it correctly.

The answer is that I haven't actually tried to train the model to convergence starting from the VGG-16 weights, so I can't tell you whether I'd be able to achieve the reference mAP values. I did train SSD300 partially multiple times to test my implementation, but when I saw that the model was learning well, I just left it at that.

The data generator I'm providing in this repository doesn't replicate the exact data augmentation procedure of the original Caffe implementation, so in order for it to make sense to try to achieve the reference mAP values when training from scratch, I would first have to implement the original data augmentation pipeline. I've so far considered this a low-priority ToDo, but maybe I'll get around to it one of these days.
